state/storage.go

```diff
@@ -701,38 +701,31 @@ func (s *Storage) RoomStateAfterEventPosition(ctx context.Context, roomIDs []str
 // - with NIDs <= `to`.
 // Up to `limit` events are chosen per room.
 func (s *Storage) LatestEventsInRooms(userID string, roomIDs []string, to int64, limit int) (map[string]*LatestEvents, error) {
-	roomIDToRanges, err := s.visibleEventNIDsBetweenForRooms(userID, roomIDs, 0, to)
+	roomIDToRange, err := s.visibleEventNIDsBetweenForRooms(userID, roomIDs, 0, to)
 	if err != nil {
 		return nil, err
 	}
 	result := make(map[string]*LatestEvents, len(roomIDs))
 	err = sqlutil.WithTransaction(s.Accumulator.db, func(txn *sqlx.Tx) error {
-		for roomID, ranges := range roomIDToRanges {
+		for roomID, r := range roomIDToRange {
 			var earliestEventNID int64
 			var latestEventNID int64
 			var roomEvents []json.RawMessage
-			// start at the most recent range as we want to return the most recent `limit` events
-			for i := len(ranges) - 1; i >= 0; i-- {
+			// the most recent event will be first
+			events, err := s.EventsTable.SelectLatestEventsBetween(txn, roomID, r[0]-1, r[1], limit)
+			if err != nil {
+				return fmt.Errorf("room %s failed to SelectEventsBetween: %s", roomID, err)
+			}
+			// keep pushing to the front so we end up with A,B,C
+			for _, ev := range events {
+				if latestEventNID == 0 { // set first time and never again
+					latestEventNID = ev.NID
+				}
+				roomEvents = append([]json.RawMessage{ev.JSON}, roomEvents...)
```

Comment:
Aside: is this performant? I would expect that this ends up making a new backing array once for each iteration.

Reply:
Yes it will do. So no, it's churny on memory. We should probably just go to the back then in-place reverse? Reducing memory spikes would be useful.

Reply:
Don't think we need to do it here. If the limit is small (~20?) then this is probably minimal?

Could also
-  `make([]json.RawMessage, len(events))` 
- write starting at index `len(events)-1` and decrease?

(Tbh I'm surprise we don't get the DB to return events in ascending NID order)

Reply:
The problem is this is done per room, so if you have a 50 window, then 50x`timeline_limit` events (worst case).

Reply:
Using `slices.Reverse` now.